File: src/main/java/com/group/libraryapp/service/user/UserService.java
package com.group.libraryapp.service.user;

import com.group.libraryapp.domain.user.JavaUser;
import com.group.libraryapp.domain.user.JavaUserRepository;
import com.group.libraryapp.dto.user.request.UserCreateRequest;
import com.group.libraryapp.dto.user.request.UserUpdateRequest;
import com.group.libraryapp.dto.user.response.UserResponse;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.stream.Collectors;

@Service
public class UserService {

  private final JavaUserRepository userRepository;

  public UserService(JavaUserRepository userRepository) {
    this.userRepository = userRepository;
  }

  @Transactional
  public void saveUser(UserCreateRequest request) {
    JavaUser newUser = new JavaUser(request.getName(), request.getAge());
    userRepository.save(newUser);
  }

  @Transactional(readOnly = true)
  public List<UserResponse> getUsers() {
    return userRepository.findAll().stream()
        .map(UserResponse::new)
        .collect(Collectors.toList());
  }

  @Transactional
  public void updateUserName(UserUpdateRequest request) {
    JavaUser user = userRepository.findById(request.getId()).orElseThrow(IllegalArgumentException::new);
    user.updateName(request.getName());
  }

  @Transactional
  public void deleteUser(String name) {
    JavaUser user = userRepository.findByName(name).orElseThrow(IllegalArgumentException::new);
    userRepository.delete(user);
  }

}

File: src/main/kotlin/com/group/libraryapp/service/user/UserService.kt
package com.group.libraryapp.service.user

import com.group.libraryapp.domain.user.User
import com.group.libraryapp.repository.user.UserRepository
import com.group.libraryapp.domain.user.loanhistory.UserLoanStatus
import com.group.libraryapp.dto.user.request.UserCreateRequest
import com.group.libraryapp.dto.user.request.UserUpdateRequest
import com.group.libraryapp.dto.user.response.BookHistoryResponse
import com.group.libraryapp.dto.user.response.UserLoanHistoryResponse
import com.group.libraryapp.dto.user.response.UserResponse
import com.group.libraryapp.util.findByIdOrThrow
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Transactional
@Service
class UserService @Autowired constructor(
    private val userRepository: UserRepository,
) {

    fun saveUser(request: UserCreateRequest) {
        val user = User(name = request.name, age = request.age)
        userRepository.save(user)
    }

    @Transactional(readOnly = true)
    fun getUsers() : List<UserResponse> {
//        return userRepository.findAll().map(UserResponse::fromEntity)
        return userRepository.findAll().map { UserResponse.fromEntity(it) }
    }

    fun updateUserName(request: UserUpdateRequest) {
        val user: User = userRepository.findByIdOrThrow(request.id)
        user.updateName(request.name)
    }

    fun deleteUser(name: String) {
        val user: User = userRepository.findByName(name) ?: throw IllegalArgumentException("해당하는 유저가 없습니다")
        userRepository.delete(user)
    }

    fun getUserLoanHistories(): List<UserLoanHistoryResponse> {
        return userRepository.findAllWithHistories().map(UserLoanHistoryResponse::of)
    }
}